Firestore/Source/API/FIRDocumentSnapshot.mm

```diff
@@ -261,41 +223,15 @@ - (id)convertedValue:(FSTFieldValue *)value options:(FSTFieldValueOptions *)opti
 
 @interface FIRQueryDocumentSnapshot ()
 
-- (instancetype)initWithFirestore:(FIRFirestore *)firestore
-                      documentKey:(DocumentKey)documentKey
-                         document:(FSTDocument *)document
-                        fromCache:(BOOL)fromCache
-                 hasPendingWrites:(BOOL)pendingWrites NS_DESIGNATED_INITIALIZER;
+- (instancetype)initWithSnapshot:(std::unique_ptr<QueryDocumentSnapshot>)snapshot
```

Comment:
This is borderline broken. I can see how it happens to work here but there are contexts in which it won't so I'm wary.

The issue is that Objective-C methods don't support overloading by type. There isn't any name mangling mechanism so there's no run-time distinction between this method and the superclass method you're overriding. This means that generally, overriding a method with a different signature can lead to some very bad behavior, especially when passing by value.

Now, couple that with the fact that the only caller of this is `+[FIRDocumentSnapshot snapshotWithSnapshot]` and you'll find out that this is not type sound: that caller accepts `std::unique_ptr<DocumentSnapshot>` and happens to be dynamically invoking this method because `[[self class] alloc]` happened to return an instance of `FIRQueryDocumentSnapshot`, but there's nothing guaranteeing the instance it passes is of this narrower type.

So, this appears to guarantee type safety, but it's really lying because you could call `+[FIRQueryDocumentSnapshot snapshotWithSnapshot:std::move(snap)]` with just a plain `std::unique_ptr<api::DocumentSnapshot>` and it would compile and run just fine. This happens to work because both types have an identical representation on the stack and we're not doing anything dependent on the type of `QueryDocumentSnapshot`, but this isn't a feature we should count on.

Separately, I think we're technically in violation of the declaration of `FIRQueryDocumentSnapshot` (and if you look closely at the build or declaration of in `FIRDocumentSnapshot.h` in Xcode you may see a warning). That declaration says that `FIRQueryDocumentSnapshot` provides its own definitions of `-data` and `-dataWithServerTimestampBehavior:` but we're not actually providing them here. This happens to work at runtime though because the instance _does_ actually have those methods, but the base class could conceivably change and this would break.

To resolve these issues:
  * Remove this initializer.
  * (maybe) add `+ (instancetype)snapshotWithSnapshot:(std::unique_ptr<QueryDocumentSnapshot>)snapshot;`. This would make calling the factory method actually type sound. However, see next points.
  * (maybe) add back the methods with different nullability here
  * (maybe) just eliminate `api::QueryDocumentSnapshot` altogether

This last point is perhaps the most interesting because it allows us to avoid making the underlying snapshot a unique pointer again and leaves the nullability fudging we're doing here as part of the language binding rather than as a part of the core API. Note that C++ does not encode nullability, so it's not clear what, if any, benefit there would be from having a public `QueryDocumentDocument` type anyway.

Reply:
> just eliminate `api::QueryDocumentSnapshot` altogether

I think this is the best approach, that type seems more trouble than it's worth.